text/0000-modular-queries.md

```diff
@@ -0,0 +1,274 @@
+
+
+- Feature Name: rfc-0000-modular-queries
+- Start Date: (fill me in with today's date, 2021-07-24)
+- Tremor Issue: [tremor-rs/tremor-runtime#940](https://github.com/tremor-rs/tremor-runtime/issues/940)
+- RFC PR: [tremor-rs/tremor-rfcs#TODO](https://github.com/tremor-rs/tremor-rfcs/pull/TODO)
+
+
+# Summary
+
+Add support for modular queries to Tremor's query language, Trickle, so that distinct subgraphs could be reused and composed into higher level queries.
+
+# Motivation
+
+Subqueries would allow composition of smaller, reusable queries into higher level queries.
+
+# Guide-level explanation
+
+## Definition
+
+![DefineSubqueryDefn](../img/rfcs/0000-modular-queries/DefineSubqueryDefn.png)
+
+```
+DefineSubqueryDefn ::= DocComment? 'define' 'subquery' Id ('from' Ports)? ('into' Ports)? WithPartialParams? Subquery
+```
+
+```
+## Documentation Comment
+define subquery custom_subquery
+from input_stream_1, input_stream_2
+into output_stream
+with
+ param1 = "foo",
+ param2 = 42
+subquery
+ select event from input_stream_1 where event.name == args.param1 into output_stream;
+ select event from input_stream_2 where event.id == args.param2 into output_stream;
+end;
+```
+
+- The `with` clause can be used to pass in parameters when required.
+    - The values given to the parameters in the with clause here act as their default values.
+    - Parameters are accessible through `args` inside the subquery.
+- The `from` clause can be used to define input streams.
+    - If elided, `in` stream is attached.
+- The `into` clause can be used to define output streams.
+    - If elided, `out` and `err` streams are attached.  
```

Comment:
Same question as for `from` above

Reply:
Yes.
Similar to `from`, if the user deviates from having just `out` **and** `err` they would need to explicitly specify the full interface.